pennylane/ftqc/lattice.py

```diff
@@ -0,0 +1,169 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# pylint: disable=protected-access, inconsistent-return-statements
+"""
+This file defines classes and functions for creating lattice objects that store topological
+connectivity information.
+"""
+
+from collections.abc import Sequence
+from enum import Enum, auto
+from functools import lru_cache
+
+import networkx as nx
+
+
+class Lattice:
+    """Represents a qubit lattice structure.
+
+    This Lattice class, inspired by the design of :class:`~pennylane.spin.Lattice`, leverages `NetworkX` to represent the relationships within the lattice structure.
+
+        Args:
+            lattice_shape: Name of the lattice shape.
+            graph (nx.Graph): A NetworkX undirected graph object. If provided, `nodes` and `edges` are ignored.
+            nodes (List): Nodes to construct a graph object. Ignored if `graph` is provided.
+            edges (List): Edges to construct the graph. Ignored if `graph` is provided.
+        Raises:
+            ValueError: If neither `graph` nor both `nodes` and `edges` are provided.
+    """
+
+    # TODOs: To support braiding operations, Lattice should support nodes/edges addition/deletion.
+
+    def __init__(
+        self, lattice_shape: str, graph: nx.Graph = None, nodes: list = None, edges: list = None
+    ):
+        self._lattice_shape = lattice_shape
+        if graph is None:
+            if nodes is None and edges is None:
+                raise ValueError(
+                    "Neither a networkx Graph object nor nodes together with edges are provided."
+                )
+            self._graph = nx.Graph()
+            self._graph.add_nodes_from(nodes)
+            self._graph.add_edges_from(edges)
+        else:
+            self._graph = graph
+
+    @property
+    def shape(self) -> str:
+        r"""Returns the lattice shape name."""
+        return self._lattice_shape
+
+    def get_neighbors(self, node):
+        r"""Returns the neighbors of a given node in the lattice.
+
+        Args:
+            node: a target node label.
+        """
+        return self._graph.neighbors(node)
+
+    @property
+    def get_nodes(self):
+        r"""Returns all nodes in the lattice."""
+        return self._graph.nodes
+
+    @property
+    def get_edges(self):
+        r"""Returns all edges in the lattice."""
+        return self._graph.edges
+
+    @property
+    def get_graph(self) -> nx.Graph:
+        r"""Returns the underlying NetworkX graph object representing the lattice."""
+        return self._graph
```

Comment:
The previous comment about `get_lattice_shape` above also goes for these properties - starting with `get_` to me implies a function that needs to be called  as `lattice.get_graph()`, whereas `lattice.graph` is a property.